### ISSUE TITLE
Custom model base breaks unions in 2.5

### ISSUE THREAD

**Issue by felix-hilden** [pydantic/pydantic#8119]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Hi! When using a custom subclass of `BaseModel`, union parsing fails altogether. I saw a potentially related issue in #8111, but there the issue was a custom raise. Is there something wrong with overriding `__init__`? I use it to provide a warning on passing extra attributes.

### Example Code

```Python
from typing import Union
from pydantic import BaseModel

content = [{"a": False}, {"b": True}]

class Model(BaseModel):
    def __init__(self, **data):
        super().__init__(**data)

class A(Model):
    a: bool

class B(Model):
    b: bool

# this is successful
A(**content[0])
B(**content[1])

class M(Model):
    m: Union[A, B]

# this fails
M(m=content[0])
M(m=content[1])
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.5.0
        pydantic-core version: 2.14.1
          pydantic-core build: profile=release pgo=true
                 install path: G:\win\dev\tekore\.venv\Lib\site-packages\pydantic
               python version: 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
             related packages: typing_extensions-4.8.0

This problem did not exist prior to 2.5
```

**Comment by sydney-runkle** [comment]
Hi @felix-hilden,

Thanks for reporting this. We're working on a fix in `pydantic-core` for this issue, and we'll get this fixed for our `2.5.2` release!

**Comment by sydney-runkle** [comment]
Alrighty! We've fixed this in `pydantic-core`, and will include this in the patch release tomorrow 😄.
